plugins/flashrom-i2c/fu-i2c-device.c

```diff
@@ -0,0 +1,288 @@
+/* Copyright 2020 The Chromium OS Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+
+#include "config.h"
+#include "fu-i2c-device.h"
+
+#include <glib.h>
+#include <glib/gstdio.h>
+
+#define LAYOUT_FLAG_NAME "FLAG"
+#define LAYOUT_PARTITION_NAME "BLK"
+#define IMG_LAYOUT_NAME "layout"
+#define IMG_FLAG1_NAME "flag1.bin"
+#define IMG_FLAG2_NAME "flag2.bin"
+#define IMG_FIRMWARE_NAME "fw.bin"
+
+struct FlashromArgs
+{
+	const gchar *spi_master;
+	const gchar *layout;
+	const gchar *image;
+	const gchar *operation;
+};
+
+struct _FuI2cDevice
+{
+	FuDevice parent_instance;
+};
+
+G_DEFINE_TYPE (FuI2cDevice, fu_i2c_device, FU_TYPE_DEVICE)
+
+/* This calls fu_common_get_contents_fd which will close the fd when done. */
+static gint fu_i2c_device_get_boot_block_from_fd (int fd, GError **error)
+{
+	g_autoptr (GBytes) result = NULL;
+	gsize result_len = 0;
+	result = fu_common_get_contents_fd (fd, 1, error);
+	if (result == NULL)
+		return -1;
+
+	const guint8 *result_bytes = g_bytes_get_data (result, &result_len);
+	if (result_len != 1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s: bootblock info size is expected to be 1 byte, \
+			     got %d",
+			     __func__,
+			     result_len);
+		return -1;
+	}
+
+	return result_bytes[0];
+}
+
+static gboolean fu_i2c_device_validate_flashrom_args (
+	const struct FlashromArgs *args,
+	GError **error)
+{
+	if (args->spi_master == NULL ||
+	    args->layout == NULL ||
+	    args->image == NULL ||
+	    args->operation == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s: all arguments under FlashromArgs has \
+			     to be set",
+			     __func__);
+		return FALSE;
+	}
+
+	return TRUE;
+
+}
+
+static gboolean fu_i2c_device_run_command (const struct FlashromArgs *args,
+					   GError **error)
+{
+	if (fu_i2c_device_validate_flashrom_args (args, error) == FALSE)
+		return FALSE;
+
+	const gchar *argv[9] = { "flashrom",
+				 "-p", args->spi_master,
+				 "--layout", args->layout,
+				 "--image", args->image,
+				 args->operation,
+				 NULL };
```

Comment:
nah, we shouldn't do this. we have a perfectly good flashrom plugin that uses libflashrom. spawning a random binary is a quick hack, sorry. You don't get errors, progress, etc.

Reply:
Yes we are aware of the flashrom plugin and ultimately we do like to use the libflashrom, however the flashrom on chromium side: https://chromium.googlesource.com/chromiumos/third_party/flashrom/ does not have the libflashrom support yet, as such feature is expected to take a long time to settle. We would like to get firmware update support now with command line flsahrom. Does it sounds reasonable?

Reply:
> however the flashrom on chromium side flashrom does not have the libflashrom support yet

Why not?

> Does it sounds reasonable?

Not really. The "spawn an executable" just hopes the flash process completes. How to you report the progress back to the user? How do we convey the various erase/write/verify stages to the front end? Do we just display a bouncing cursor back and forth for 5 minutes? 30 seconds? What if it hangs during initialisation?

Reply:
I think this turns into a bit of a larger question - what is actually different with that flashrom and https://review.coreboot.org/plugins/gitiles/flashrom?  Is the chromium one a fork or just moving at a faster pace?  Or  are things actually flowing back up to the the one at coreboot?

If you take this Parade case, I would wonder specifically if there needs to be specific code on the flashrom side only present in your fork to handle it, or if the (recently tagged) flashrom 1.2 has it?  What kind of check could be done at compile or even runtime time to know there is support?

>as such feature is expected to take a long time to settle

Have they really deviated so much that you can't cherry pick the 25 commits into your fork for it?